### ISSUE TITLE
Generalize optimization like log(sigmoid) or log(1-sigmoid)  to tolerate reshape/dimshuffle

### ISSUE THREAD

**Issue by nouiz** [Theano/Theano#4451]
We should build helper function to we could reuse as they cause "problem" at a few different place.

One report:

https://groups.google.com/d/topic/theano-users/i1kKZmclFFo/discussion

**Comment by JesseLivezey** [comment]
Also, #3254 (transpose)

@nouiz, where do the current optimizations live? I looked in tensor.nnet.sigm, but couldn't find them, although I'm not sure what they would be named.

Would this help with log(softmax) also?
log(softmax) seems to be here
https://github.com/Theano/Theano/blob/master/theano/tensor/nnet/nnet.py#L740
crossentropy(softmax) here
https://github.com/Theano/Theano/blob/master/theano/tensor/nnet/nnet.py#L1662

This is somewhat unrelated, but I also remember some discussion about whether crossentropy(softmax) was being optimized for stability when the target distribution was a vector rather than an index. Can't find the discussion, but I might be interested in adding this also.

**Comment by nouiz** [comment]
log(sidmoid) is here:

https://github.com/Theano/Theano/blob/master/theano/tensor/nnet/sigm.py#L383

**Comment by JesseLivezey** [comment]
Something like
```
logsigm_to_softplus = gof.PatternSub(
    (tensor.log, (reshape_dimshuffle_transpose, (sigmoid, 'x'))),
    (reshape_dimshuffle_transpose, (tensor.neg, (softplus, (tensor.neg, 'x')))),
    allow_multiple_clients=True,
    values_eq_approx=values_eq_approx_remove_inf,
    skip_identities_fn=_skip_mul_1)
```
where ```reshape_dimshuffle_transpose``` is any combination of the three or nothing (original case).

**Comment by nouiz** [comment]
Yes. Transpose are dimshuffle, you you only need to check for
reshape/dimshuffle.

**Comment by f0k** [comment]
> This is somewhat unrelated, but I also remember some discussion about whether crossentropy(softmax) was being optimized for stability when the target distribution was a vector rather than an index. Can't find the discussion, but I might be interested in adding this also.

https://github.com/Theano/Theano/issues/3162#issuecomment-123123283. Seems to have been resolved in #3363?

**Comment by nouiz** [comment]
Yes, it fix the numerical stabilisation problem.

This PR is to make this optimization applied more often.

**Comment by nouiz** [comment]
See https://github.com/Theano/Theano/issues/4641 for an alternative way to fix this problem that I think is more generic.

**Comment by JesseLivezey** [comment]
Great, something like that is what I was thinking, but it would be good to make it generic to elemwise.

I can put together a draft this weekend.

**Comment by nouiz** [comment]
If you want to work on the reshape lifters it will be useful. Yes I was
thinking of lifting reshape just on elemwise like what we do for dimshuffle.

Le ven. 17 juin 2016 13:21, Jesse Livezey <notifications@github.com> a
écrit :
